format.md

```diff
@@ -130,6 +130,64 @@ is a list of xy vertices. The child name of the outer list should be "polygons";
 the child name of the middle list should be "rings"; the child name of the
 inner list should be "vertices".
 
+**Geometry**: `DenseUnion`
+
+So far, all geometry array types listed above have required that all geometries
+in the array be of the same type. An array of mixed geometry type is represented
+as a [dense
+union](https://arrow.apache.org/docs/format/Columnar.html#dense-union) whose
+children include one or more of the above geometry array types.
+
+The geometry array allows for elements in the array to be of different geometry types.
+
+- The union array may not contain more than one child array of a given geometry type.
+- The "type ids" and field name of the union field metadata must be defined as such:
```

Comment:
Given that we require that there is only one dimensionality (i.e all XY, or all XYZ, etc) for all types, do we need those dimensionality-specific type ids and names? 

We could also "just" have the 6 basic types, without any indication of the dimension (that follows from the memory layout itself, just as it is for the other geometry types. It seems a bit inconsistent to here have specific type names)

(note: this seems simpler to me, but that feeling is not based on actually implementing it ;))

Reply:
> Given that we require that there is only one dimensionality (i.e all XY, or all XYZ, etc) for all types, do we need those dimensionality-specific type ids and names?

Is there any possibility that we might want to allow multiple dimensionalities in the future?

I don't think it's much difference in implementing work whether we use double-digit type ids for XYZ or not, and I don't have a strong preference.

Reply:
To be clear, I wasn't arguing against type ids in general, only about having separate type ids for > 2D geometry types, i.e. all types ids > 6.

> Is there any possibility that we might want to allow multiple dimensionalities in the future?

Potentially yes, but that's also something that could still be expanded in the future even if we only have type ids 1-6 now?

Reply:
I'm very pro on _having type ids_. Having standardized type ids made it a lot easier to fix some type/field issues in geoarrow-rs.

I read @jorisvandenbossche 's comment as saying we could mandate:

| Type ID | Geometry type      | Field name             |
| ------- | ------------------ | ---------------------- |
| 1       | Point              | `"Point"`              |
| 2       | LineString         | `"LineString"`         |
| 3       | Polygon            | `"Polygon"`            |
| 4       | MultiPoint         | `"MultiPoint"`         |
| 5       | MultiLineString    | `"MultiLineString"`    |
| 6       | MultiPolygon       | `"MultiPolygon"`       |

and have all dimensions of point arrays use type id `1`

Reply:
Indeed, and in that case I would maybe bikeshed on the Field name, that this could be lower-case to match the names we use for the single-geometry-type extension types (the `"geoarrow.<type>"` is using "point", "linestring" et al. Although that is purely metadata of course, while here it is part of the type)

Reply:
> unless there is any future where POINT and POINT Z will be allowed in the same array, in which case we'd need a new type rather than relaxing a restriction on an existing type

Right. I'm still slightly leaning towards having different type ids per dimension so that it would be _possible_ to relax this restriction in the future if we so desired.

Reply:
That sounds sensible yes. But I think that also means we can add type ids>10 later if needed? (if old consumers validate that only type ids 1-7 are present)

Reply:
I agree that unions can be painful, but it's still useful that the array _isn't_ serialized, so 95% of the time it can be cast directly to one of its constituent parts, without any copies.